### PR TITLE
Add support for changing the final boss

### LINKED ISSUE #428: option to change finalboss
make ridley not end game, but instead some other boss

### PR DESCRIPTION
**Pull request by dyceron**
Okay, so this PR is kinda wild. I've imagined this idea for a while and decided to take a stab at it. It's not very traditional, but it's yet another way to spice up the gameplay up a little bit. With this, any of the four bosses (Arachnus, Diggernaut, Queen, Ridley) can be the final boss! To summarize the changes I made:

- I moved `required_dna` to a new field called `objective` which also lists the boss required to finish.
- Beating the configured boss completes the game and rolls the credits
- Just like with accessing Ridley, accessing the new final boss is also locked to DNA
  - If the player doesn't have enough DNA, a message will appear on screen. Arachnus and Diggernaut will warp Samus back to the previous cc
- If Ridley is not the final boss:
  - Access to the fight is always active until Ridley is dead, and a textbox prompt is shown asking the player if they want to proceed or continue to the ship
  - After Ridley is dead, Samus will respawn at the ship and a new pickup is added to the right of the ship. The Ridley landing site is also permanently inaccessible
- If Arachnus is the final boss:
  - The credits will not roll until the exterior room is re-entered to prevent memory errors
  - The pickup will not spawn at all
  - Arachnus' health is buffed and beams/Screw Attack is nerfed. Power Bombs no longer work
    - Debating on if it's worth also buffing Arachnus's speed/length of attacks, if that's even possible
- If Diggernaut is the final boss:
  - The credits roll after the death cutscene
  - The grapple block near the elevator is always moved regardless if `remove_elevator_grapple_blocks` is enabled or not to allow access to the elevator
- If the Queen is the final boss:
  - The credits roll after the death cutscene
  - The wall behind the Queen is always removed regardless if `reverse_area8` is enabled to allow for traversal

Some screenshots to showcase this option

Arachnus Final Boss
image

Diggernaut Final Boss
image

Queen Final Boss
image

Ridley Not Final Boss
image

Fixes #428

### CODE REVIEW

**Comment by ThanatosGit** [comment]
I just read the description here so far: Is there any way to make it multiworld compatible in RDV? I have no idea how you could prevent the generator from placing other world's DNA - or whatever macguffin - behind e.g. Diggernaut. But me having no clue how you would do something related to logic is not uncommon :D

**Comment by dyceron** [comment]
I believe it's possible with DB layers and/or bootstrap? I don't know the full details of that just yet, but it would be nice to have this MW compatible if possible.

**Comment by steven11sjf** [comment]
You can definitely make arachnus deal more damage via tunables. it might be possible to alter attack speeds by messing with BMSAS (iirc part of BMSAD) and possibly BCSKLA, moving keyframes around. 

Overall I'm not really sure how to feel about Arachnus being a final boss. I definitely think queen should be an option, and I see the case for diggernaut since it's a mid-late game boss, but an A2 boss seems pretty underpowered and turning it into a damage sponge feels lazy.

**Comment by Miepee** [comment]
I'm pretty sure this feature was inspired by VARIA where you can customize the goal to beat any of the existing (mini)bosses. 
I do think that turning Arachnus into a bullet sponge is overkill, but this PR also doesnt really do that to my knowledge. It only buffs Arachnus to not die faster than you can say "Hi Arachnus" with like 70% of the items in the game. Aka, make him an actual miniboss instead of a glorified enemy.

**Comment by dyceron** [comment]
I did consider not including Arachnus, bur for completeness sake I wanted it. More tweaks for it will make it more worth it.

File: src/open_samus_returns_rando/files/levels/s110_surfaceb.lua [hunk 1]
```diff
@@ -299,6 +308,11 @@ function s110_surfaceb.LaunchRidleyDeadCutscene()
   GUI.ForceIdleCinematicControlledOn()
   Game.LaunchCutscene("cutscenes/ridley4/takes/10/ridley410.bmscu")
   s110_surfaceb.SetLowModelsVisibility(false)
+  if Init.sFinalBoss ~= "Ridley" then
+    Blackboard.SetProp("DEFEATED_ENEMIES", "Ridley", "b", true)
```
Comment (ThanatosGit) [inline src/open_samus_returns_rando/files/levels/s110_surfaceb.lua:312]: I cannot find a `SetProp` for the other enemies: I assume they are already set by game internal stuff?

File: src/open_samus_returns_rando/files/levels/s020_area2.lua [hunk 2]
```diff
@@ -64,7 +68,17 @@ end
 function s020_area2.OnExit()
 end
 function s020_area2.OnSubAreaChange(_ARG_0_, _ARG_1_, _ARG_2_, _ARG_3_, _ARG_4_)
+  if _ARG_0_ == "" and _ARG_2_ == "collision_camera_005" and Init.sFinalBoss == "Arachnus" and 0 < Scenario.ReadFromBlackboard("entity_SG_Arachnus_001_deaths", 0) then
+    s020_area2.bAfterCredits = false
```
[the remark this answers is not in the document]
Comment (dyceron) [reply src/open_samus_returns_rando/files/levels/s020_area2.lua:72]: This is just a copy paste from SurfaceB. I didn't test without it so I left it.

File: src/open_samus_returns_rando/misc_patches/final_boss.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    if final_boss != "Ridley":
+        # Add new boss triggers
+        for new_trigger in new_triggers:
+            add_boss_triggers(editor, new_trigger)
+        if final_boss == "Arachnus":
+            # Buff Arachnus so the fight isn't trivial
+            arachnus = editor.get_file("actors/characters/arachnus/charclasses/arachnus.bmsad", Bmsad)
+            bmsad_life = arachnus.raw["components"]["LIFE"]["fields"]
+            # Disable Power Bombs because of the instant kill
+            bmsad_life["bShouldDieWithPowerBomb"] = Container({"type": "bool", "value": False})
```
[the remark this answers is not in the document]
Comment (dyceron) [reply src/open_samus_returns_rando/misc_patches/final_boss.py:53]: I remember testing this a while ago and it worked then. I think Diggernaut is just special or something.

File: src/open_samus_returns_rando/files/levels/s110_surfaceb.lua (hunk 1 above)
Comment (dyceron) [reply src/open_samus_returns_rando/files/levels/s110_surfaceb.lua:312]: It depends. Metroids check for a deaths value, the Queen sets a prop, so I did the latter here.

File: src/open_samus_returns_rando/files/levels/s000_surface.lua [hunk 4]
```diff
@@ -93,6 +93,11 @@ function s000_surface.InitFromBlackboard()
     Game.SetPlayerInputEnabled(false, false)
     s000_surface.LaunchPlanetArrival()
   end
+  if not Blackboard.GetProp("DEFEATED_ENEMIES", "Ridley") and Game.GetEntity("LE_Item_Ridley") ~= nil then
+    Game.DisableEntity("LE_Item_Ridley")
+  else
+    Game.EnableEntity("LE_Item_Ridley")
+  end
```
[the remark this answers is not in the document]
Comment (dyceron) [reply src/open_samus_returns_rando/files/levels/s000_surface.lua:100]: I feel it shouldn't respawn? I'll need to test this.

**Comment by Miepee** [comment]
Out of curiosity, is there a reason why buffed arachnus is only available when he's the final boss? Feel like general rando could also benefit from him not being a complete pushover

**Comment by dyceron** [comment]
> Out of curiosity, is there a reason why buffed arachnus is only available when he's the final boss? Feel like general rando could also benefit from him not being a complete pushover

Feels out of place to adjust in general. As the final boss it makes sense otherwise it's a complete joke.

**Comment by Miepee** [comment]
IDK, balancing arachnus to not die in one hit seems like in-line for me with balancing diggernaut and ridley.

**Comment by dyceron** [comment]
I don't want to make arachnus crazy powerful considering it can be an early check normally. Regardless, that's out of scope for this pr.

File: src/open_samus_returns_rando/files/levels/s000_surface.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (dyceron) [reply src/open_samus_returns_rando/files/levels/s000_surface.lua:100]: Confirmed it does not respawn

**Comment by dyceron** [comment]
Fixed some stuff and removed unused stuff

File: src/open_samus_returns_rando/files/levels/s000_surface.lua (hunk 4 above)
Comment (ThanatosGit) [inline src/open_samus_returns_rando/files/levels/s000_surface.lua:100]: There was something weird about this check which I forgot about: Can move and surround both by the `Game.GetEntity("LE_Item_Ridley") ~= nil` check? It kinda scares me to call `EnableEntity` on a non existing one.

**Comment by dyceron** [comment]
Made some adjustments to starting DNA and also change the message when collecting all DNA for each boss. I wasn't sure if I wanted to just not display the Baby hint if the final boss isn't Ridley, so I left it in.

All DNA and Baby

image

All DNA and no Baby

image

File: src/open_samus_returns_rando/files/custom/scenario.lua [hunk 5]
```diff
@@ -201,6 +201,39 @@ function Scenario.RandoOnElevatorUse(from_actor, _ARG_1_, _ARG_2_)
...
+function Scenario.LaunchCredits()
+  Game.ShowEndGameCredits(true)
+end
+
+function Scenario.OnPostCreditsEnd()
+  Game.SaveGameComplete()
+  if CurrentScenario.bFirstTimeCompleted then
+    Game.StopEnvironmentSound()
+    Game.AddGUISF(2.5, GUI.MainMenuGoToState, "i", 28)
+  else
```
Comment (ThanatosGit) [inline src/open_samus_returns_rando/files/custom/scenario.lua:232]: Does this have any use now? Before the `bFirstTimeCompleted` was added to every scenario but was always false and now it doesn't even exist besides the vanilla thing in s110_surfaceb.

Comment (dyceron) [reply src/open_samus_returns_rando/files/custom/scenario.lua:232]: Honestly I'm not even sure. It seems to always just do the else regardless. I even deleted all Save data and hard was just unlocked on the file and completed after finishing, but had no popup. Probably could be removed then

Comment (dyceron) [reply src/open_samus_returns_rando/files/custom/scenario.lua:232]: Even when the scenarios had those variables it still was never called so I'm not sure what it's for. I think Ridley is handled differently.

Comment (dyceron) [reply src/open_samus_returns_rando/files/custom/scenario.lua]: Tested without it and it still finished just fine, so I'm removing it. Hopefully that's everything?

**Review by ThanatosGit — APPROVED** [review]

**Pull request merged**